src/rgw/rgw_sal_d4n.cc

```diff
@@ -0,0 +1,144 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "rgw_sal_d4n.h"
+
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
+
+namespace rgw { namespace sal {
+
+static inline Object* nextObject(Object* t)
+{
+  if (!t)
+    return nullptr;
+  
+  return dynamic_cast<FilterObject*>(t)->get_next();
+}  
+
+int D4NFilterStore::initialize(CephContext *cct, const DoutPrefixProvider *dpp)
+{
+  FilterStore::initialize(cct, dpp);
+  blk_dir->init(cct);
+  
+  return 0;
+}
+
+std::unique_ptr<Object> D4NFilterStore::get_object(const rgw_obj_key& k)
+{
+  std::unique_ptr<Object> o = next->get_object(k);
+  return std::make_unique<D4NFilterObject>(std::move(o), this);
+}
+
+std::unique_ptr<Writer> D4NFilterStore::get_atomic_writer(const DoutPrefixProvider *dpp,
+				  optional_yield y,
+				  std::unique_ptr<rgw::sal::Object> _head_obj,
+				  const rgw_user& owner,
+				  const rgw_placement_rule *ptail_placement_rule,
+				  uint64_t olh_epoch,
+				  const std::string& unique_tag)
+{
+  /* We're going to lose _head_obj here, but we don't use it, so I think it's
+   * okay */
+  std::unique_ptr<Object> no = nextObject(_head_obj.get())->clone();
+
+  std::unique_ptr<Writer> writer = next->get_atomic_writer(dpp, y, std::move(no),
+							   owner, ptail_placement_rule,
+							   olh_epoch, unique_tag);
+
+  return std::make_unique<D4NFilterWriter>(std::move(writer), this, std::move(_head_obj), dpp);
+}
+
+std::unique_ptr<Object::ReadOp> D4NFilterObject::get_read_op()
+{
+  std::unique_ptr<ReadOp> r = next->get_read_op();
+  return std::make_unique<D4NFilterReadOp>(std::move(r), this);
+}
+
+std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+{
+  std::unique_ptr<DeleteOp> d = next->get_delete_op();
+  return std::make_unique<D4NFilterDeleteOp>(std::move(d), this);
+}
+
+int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
+{
+  int getReturn = source->filter->get_block_dir()->getValue(source->filter->get_cache_block());
+
+  if (getReturn < 0) {
+    ldpp_dout(dpp, 20) << "D4N Filter: Directory get operation failed." << dendl;
```

Comment:
What logging level is best in this case?

Matt's response: "if this even should 'never happen' (it's a serious error) then level 0. If it is relatively normal (we tried to get something but it doesn't happen to be cached), some level from 7 up to 20, lower if this is something basic that would be useful to a large number of debugging scenarios."